In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
from tensorflow import keras

In [3]:
df=pd.read_csv('Iris_dataset.csv')

In [4]:
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [5]:
df['Species'].value_counts()

Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: Species, dtype: int64

In [6]:
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

In [7]:
Y=pd.get_dummies(df.Species)

In [8]:
Y

Iris-setosa  Iris-versicolor  Iris-virginica
0              1                0               0
1              1                0               0
2              1                0               0
3              1                0               0
4              1                0               0
..           ...              ...             ...
145            0                0               1
146            0                0               1
147            0                0               1
148            0                0               1
149            0                0               1

[150 rows x 3 columns]

In [9]:
X=df.drop(columns=['Id','Species'])
X.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0            5.1           3.5            1.4           0.2
1            4.9           3.0            1.4           0.2
2            4.7           3.2            1.3           0.2
3            4.6           3.1            1.5           0.2
4            5.0           3.6            1.4           0.2

In [25]:
Y

Iris-setosa  Iris-versicolor  Iris-virginica
0              1                0               0
1              1                0               0
2              1                0               0
3              1                0               0
4              1                0               0
..           ...              ...             ...
145            0                0               1
146            0                0               1
147            0                0               1
148            0                0               1
149            0                0               1

[150 rows x 3 columns]

In [11]:
x_train,x_test,y_train,y_test=train_test_split(X,Y,test_size=0.25,shuffle=True,random_state=69)

In [12]:
x_train

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
50             7.0           3.2            4.7           1.4
149            5.9           3.0            5.1           1.8
15             5.7           4.4            1.5           0.4
115            6.4           3.2            5.3           2.3
84             5.4           3.0            4.5           1.5
..             ...           ...            ...           ...
22             4.6           3.6            1.0           0.2
49             5.0           3.3            1.4           0.2
90             5.5           2.6            4.4           1.2
107            7.3           2.9            6.3           1.8
54             6.5           2.8            4.6           1.5

[112 rows x 4 columns]

In [13]:
y_train

Iris-setosa  Iris-versicolor  Iris-virginica
50             0                1               0
149            0                0               1
15             1                0               0
115            0                0               1
84             0                1               0
..           ...              ...             ...
22             1                0               0
49             1                0               0
90             0                1               0
107            0                0               1
54             0                1               0

[112 rows x 3 columns]

In [14]:
from keras import Sequential
from keras.layers import Dense,Dropout

In [15]:
model=Sequential()
model.add(Dense(units=32,activation="relu"))
model.add(Dense(units=32,activation="relu"))
model.add(Dense(units=3,activation="softmax"))


In [20]:
model.compile(optimizer='Adam',loss='categorical_crossentropy',metrics=['accuracy'])
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 32)                  │             160 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           1,056 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 3)                   │              99 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 1,315 (5.14 KB)

 Trainable params: 1,315 (5.14 KB)

 Non-trainable params: 0 (0.00 B)

In [21]:
model.fit(x_train,y_train, epochs=100 ,batch_size=5)

Epoch 1/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9569 - loss: 0.0704   
Epoch 2/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9801 - loss: 0.0604 
Epoch 3/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9720 - loss: 0.0845 
Epoch 4/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9808 - loss: 0.0598 
Epoch 5/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9667 - loss: 0.0920 
Epoch 6/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9691 - loss: 0.0589 
Epoch 7/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9332 - loss: 0.1351 
Epoch 8/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9735 - loss: 0.1089 
Epoch 9/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9802 - loss: 0.0838 
Epoch 10/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9587 - loss: 0.0928 
Epoch 11/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9698 - loss: 0.0536 
Epoch 12/100
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/ste

In [24]:
prediction=model.predict(x_test)
print(prediction)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
[[9.9999976e-01 1.8361480e-07 9.4817199e-13]
 [3.8264834e-06 9.9971932e-01 2.7689055e-04]
 [9.9999869e-01 1.3040724e-06 3.3088105e-12]
 [8.6639397e-11 7.6357828e-04 9.9923646e-01]
 [9.9999893e-01 1.1207360e-06 4.4613926e-12]
 [1.0000000e+00 7.7498790e-09 1.3783291e-14]
 [9.6373606e-06 9.9588722e-01 4.1031209e-03]
 [8.7300785e-11 2.5589827e-03 9.9744105e-01]
 [6.6270886e-08 6.4566773e-01 3.5433218e-01]
 [1.0000000e+00 2.1418112e-08 4.9035371e-14]
 [5.4414194e-08 8.0328278e-02 9.1967171e-01]
 [2.4747251e-12 4.1983274e-04 9.9958020e-01]
 [9.9999225e-01 7.7212298e-06 1.4078751e-10]
 [3.6608060e-06 9.9575728e-01 4.2390428e-03]
 [7.5602397e-10 5.0846138e-03 9.9491537e-01]
 [3.2404000e-06 9.9777150e-01 2.2253417e-03]
 [5.6182518e-05 9.9986112e-01 8.2695740e-05]
 [1.6693010e-11 5.1351381e-04 9.9948645e-01]
 [1.3369981e-05 9.9971718e-01 2.6937868e-04]
 [5.1803628e-10 1.7095006e-03 9.9829048e-01]
 [2.4848760e-12 2.2879509e-04 9.9977118e-01]
 [2.2216427e-05 9

In [23]:
ans=model.evaluate(x_test,y_test)
ans

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9720 - loss: 0.0438  


[0.04547794163227081, 0.9736841917037964]

## Hence we have made a model with 97% accuracy which is able to classify the species of flower as [1,0,0] for iris_setosa [0,1,0] for iris_versicolor and [0,0,1] for iris_virginica 